File: testbed/test/cruise/associations/OptionalOneToManyTest.java
package cruise.associations;

import org.junit.*;

import cruise.associations.MentorC;
import cruise.associations.StudentC;

public class OptionalOneToManyTest
{

  @Test
  public void CreateStudentWithoutMentor()
  {
    StudentC s = new StudentC(99);
    Assert.assertEquals(null,s.getMentor());
  }
  
  @Test
  public void addStudent()
  {
    MentorC m = new MentorC("m1");
    StudentC s = new StudentC(99);
    m.addStudent(s);
    
    Assert.assertEquals(1, m.numberOfStudents());
    Assert.assertEquals(m, s.getMentor());
    Assert.assertEquals(s, m.getStudent(0));
  }
  
  @Test
  public void replaceMentor()
  {
    MentorC m = new MentorC("m1");
    MentorC m2 = new MentorC("m2");

    StudentC s = new StudentC(99);
    m.addStudent(s);
    Assert.assertEquals(1, m.numberOfStudents());

    
    s.setMentor(m2);
    
    Assert.assertEquals(m2, s.getMentor());
    Assert.assertEquals(s, m2.getStudent(0));
    Assert.assertEquals(0, m.numberOfStudents());
  } 
  
  @Test
  public void addToNewMentor()
  {
    MentorC m = new MentorC("m1");
    MentorC m2 = new MentorC("m2");
    StudentC s = new StudentC(99);
    
    m.addStudent(s);
    m2.addStudent(s);
    
    Assert.assertEquals(m2, s.getMentor());
    Assert.assertEquals(s, m2.getStudent(0));
    Assert.assertEquals(0, m.numberOfStudents());
  }  
  
  @Test
  public void removeFromExistingMentor()
  {
    MentorC m = new MentorC("m1");
    MentorC m2 = new MentorC("m2");

    StudentC s1 = new StudentC(99);
    StudentC s2 = new StudentC(98);
    StudentC s3 = new StudentC(97);
    
    m.addStudent(s1);
    m.addStudent(s2);
    m.addStudent(s3);

    Assert.assertEquals(3, m.numberOfStudents());
    Assert.assertEquals(0, m2.numberOfStudents());
    
    m2.addStudent(s1);

    Assert.assertEquals(2, m.numberOfStudents());
    Assert.assertEquals(s2, m.getStudent(0));
    Assert.assertEquals(s3, m.getStudent(1));
    
    Assert.assertEquals(1, m2.numberOfStudents());
    Assert.assertEquals(s1, m2.getStudent(0));
    
    Assert.assertEquals(m2, s1.getMentor());
    Assert.assertEquals(m, s2.getMentor());
    Assert.assertEquals(m, s3.getMentor());
  }   
  
  @Test
  public void removeStudent()
  {
    MentorC m = new MentorC("m1");
    StudentC s = new StudentC(99);
    
    m.addStudent(s);
    m.removeStudent(s);
    
    Assert.assertEquals(null,s.getMentor());
    Assert.assertEquals(0,m.numberOfStudents());
    
  }  

  
  
}

File: testbed_php/test/associations/OptionalOneToManyTest.php
<?php

class OptionalOneToManyTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentC(99);
    $this->assertEqual(null,$s->getMentor());
  }
  

  public function test_addStudent()
  {
    $m = new MentorC("m1");
    $s = new StudentC(99);
    $m->addStudent($s);
    
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual($s,$m->getStudent_index(0));
  }
  

  public function test_replaceMentor()
  {
    $m = new MentorC("m1");
    $m2 = new MentorC("m2");

    $s = new StudentC(99);
    $m->addStudent($s);
    $this->assertEqual(1,$m->numberOfStudents());
 
    $s->setMentor($m2);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent_index(0));
    $this->assertEqual(0,$m->numberOfStudents());
  } 
  

  public function test_addToNewMentor()
  {
    $m = new MentorC("m1");
    $m2 = new MentorC("m2");
    $s = new StudentC(99);
    
    $m->addStudent($s);
    $m2->addStudent($s);
    
    $this->assertEqual($m2,$s->getMentor());
    $this->assertEqual($s,$m2->getStudent_index(0));
    $this->assertEqual(0,$m->numberOfStudents());
  }  
  

  public function test_removeFromExistingMentor()
  {
    $m = new MentorC("m1");
    $m2 = new MentorC("m2");

    $s1 = new StudentC(99);
    $s2 = new StudentC(98);
    $s3 = new StudentC(97);
    
    $m->addStudent($s1);
    $m->addStudent($s2);
    $m->addStudent($s3);

    $this->assertEqual(3,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    
    $m2->addStudent($s1);

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($s2,$m->getStudent_index(0));
    $this->assertEqual($s3,$m->getStudent_index(1));
    
    $this->assertEqual(1,$m2->numberOfStudents());
    $this->assertEqual($s1,$m2->getStudent_index(0));
    
    $this->assertEqual($m2,$s1->getMentor());
    $this->assertEqual($m,$s2->getMentor());
    $this->assertEqual($m,$s3->getMentor());
  }   
  

  public function test_removeStudent()
  {
    $m = new MentorC("m1");
    $s = new StudentC(99);
    
    $m->addStudent($s);
    $m->removeStudent($s);
    
    $this->assertEqual(null,$s->getMentor());
    $this->assertEqual(0,$m->numberOfStudents());
    
  }  


  public function test_delete()
  {
    $m = new MentorC("m1");
    $s = new StudentC(99);

    $m->addStudent($s);

    $m->delete();
    $this->assertNull($s->getMentor());
    $this->assertFalse($m->hasStudents());
  }

  
  public function test_deleteSubclass()
  {
    $m = new MentorCSub("m1");
    $s = new StudentC(99);

    $m->addStudent($s);

    $m->delete();
    $this->assertNull($s->getMentor());
    $this->assertFalse($m->hasStudents());
  } 
  
}

?>
